src/main/java/org/mastodon/views/trackscheme/display/PaintHierarchicalGraph.java

```diff
@@ -0,0 +1,156 @@
+	}
+
+	private void calculateAverageLetterWidth()
+	{
+		final FontRenderContext frc = g2.getFontRenderContext();
+		String text = "0123456789abcdefghijklmnopqrstuvwxyz";
+		final TextLayout layout = new TextLayout( text, style.getFont(), frc );
+		averageLetterWidth = layout.getBounds().getWidth() / text.length();
+	}
+
+	@Override
+	protected void drawVertex( ScreenVertex vertex )
+	{
+		if(spotRadius > 3)
+			drawVertexFull( vertex );
+		else
+			drawVertexSimplified( vertex );
+	}
+
+	@Override
+	protected void drawVertexFull( ScreenVertex vertex )
+	{
+		final Transition transition = vertex.getTransition();
+		final boolean disappear = ( transition == DISAPPEAR );
+		final double ratio = vertex.getInterpolationCompletionRatio();
+
+		final boolean highlighted = ( highlightedVertexId >= 0 ) && ( vertex.getTrackSchemeVertexId() == highlightedVertexId );
+		final boolean focused = ( focusedVertexId >= 0 ) && ( vertex.getTrackSchemeVertexId() == focusedVertexId );
+		final boolean selected = vertex.isSelected();
+		final boolean ghost = vertex.isGhost();
+		final int specifiedColor = vertex.getColor();
+
+		final Color fillColor = getColor( selected, ghost, transition, ratio, specifiedColor,
+				style.getVertexFillColor(), style.getSelectedVertexFillColor(),
+				style.getGhostVertexFillColor(), style.getGhostSelectedVertexFillColor() );
+		final Color drawColor = getColor( selected, ghost, transition, ratio, 0,
+				style.getVertexDrawColor(), style.getSelectedVertexDrawColor(),
+				style.getGhostVertexDrawColor(), style.getGhostSelectedVertexDrawColor() );
+
+		final double x = vertex.getX();
+		final double y = vertex.getY();
+		final int ox = ( int ) x - ( int ) spotRadius;
+		final int oy = ( int ) y - ( int ) spotRadius;
+		final int sd = 2 * ( int ) spotRadius;
+		g2.setColor( fillColor );
+		g2.fillOval( ox, oy, sd, sd );
+
+		g2.setColor( drawColor );
+		if ( highlighted )
+			g2.setStroke( style.getVertexHighlightStroke() );
+		else if ( focused )
+			// An animation might be better for the focus, but for now this is it.
+			g2.setStroke( style.getFocusStroke() );
+		else if ( ghost )
+			g2.setStroke( style.getVertexGhostStroke() );
+		g2.drawOval( ox, oy, sd, sd );
+		if ( highlighted || focused || ghost )
+			beforeDrawVertices();
+
+		final int maxLabelLength = (int) (transform.getScaleX() * 0.8 / averageLetterWidth );
+		if ( maxLabelLength > 2 && !disappear )
+		{
+			String label = vertex.getLabel();
+			if ( label.length() > maxLabelLength )
+				label = label.substring( 0, maxLabelLength - 2 ) + "...";
+
+			if ( !label.isEmpty() )
+			{
+				g2.setColor( labelColor );
+				final FontRenderContext frc = g2.getFontRenderContext();
+				final TextLayout layout = new TextLayout( label, style.getFont(), frc );
+				final Rectangle2D bounds = layout.getBounds();
+				final float tx = ( float ) ( x - bounds.getCenterX() );
+				final float ty = ( float ) ( y + bounds.getHeight() + 1.2 * spotRadius + STROKE_WIDTH );
```

Comment:
With this the round circle of the node is a bit empty and feels useless. 
Would you feel like make a display mode where the labels are painted in the nodes, but with nodes having a rounded rectangle shape? Its max width would be set by the column width of the branch child breadth.
Then we could make a flag or an enum that could let the user decide how they want to paint them.
What do you think?

Reply:
Hehe :) OK.